File: longest_k_unqiue_characters_substring.cpp
//{ Driver Code Starts
//Initial template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function template for C++

class Solution{
  public:
    int longestKSubstr(string s, int k) {
    // your code here
        int ans = -1;
        int j = 0;
        int used = 0;
        int c[256] = {0};
        for(int i = 0; i < s.size(); ++i) {
            if (c[s[i]]++ == 0) {
                used++;
            }
            if (used == k) {
                ans = max(i-j+1, ans);
            } else if (used > k) {
                while (--c[s[j]] != 0) {
                    ++j;
                }
                --used;
                ++j;
                ans = max(i-j+1, ans);
            }
        }
        return ans;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        string s;
        cin >> s;
        int k;
        cin >> k;
        Solution ob;
        cout << ob.longestKSubstr(s, k) << endl;
    }
}

// } Driver Code Ends

File: two_pointers/longest_k_unqiue_characters_substring.py
# User function Template for python3


class Solution:
    def longestKSubstr(self, s, k):
        # code here
        j = 0
        n = len(s)
        ans = -1
        cnt = [0 for _ in range(26)]
        for i in range(n):
            d = ord(s[i]) - ord("a")
            cnt[d] += 1
            if cnt[d] == 1:
                k -= 1

            while k < 0:
                cnt[ord(s[j]) - ord("a")] -= 1
                if cnt[ord(s[j]) - ord("a")] == 0:
                    k += 1
                j += 1

            if k == 0:
                ans = max(ans, i - j + 1)

        return ans


# {
# Driver Code Starts
# Initial Template for Python 3

if __name__ == "__main__":
    t = int(input())

    for _ in range(t):
        s = input()
        k = int(input())

        solObj = Solution()

        ans = solObj.longestKSubstr(s, k)

        print(ans)

# } Driver Code Ends
